### PR TITLE
PAAS-3674 allow updating matchlabels when users know about manual cle…

### PR DESCRIPTION
**Pull request by grosser**
…anup

@zendesk/compute

### CODE REVIEW

**Review by vvicaretti — APPROVED** [review]
👍

File: plugins/kubernetes/README.md [hunk 1]
```diff
@@ -247,3 +247,14 @@ To debug services or to create resources that needs to reference a selector that
...
+### Updating matchLabels
+
+Samson will by default block updating `matchLabels` since it leads to abandoned pods.
+
+If you still want to change a matchLabel, for example project/role:
+
+ - set `metadata.annotations.samson/allow_updating_match_labels: "true"`
+ - deploy with renamed project/role
+ - manually delete pods from old Deployment (`kubectl delete pods -l project=old-name,role=old-role`)
+ - unset annotations from step 1
```
Comment (adammw) [inline plugins/kubernetes/README.md]: finda kunky.. would prefer a checkbox on the deploy, but `¯\_(ツ)_/¯`

File: plugins/kubernetes/app/models/kubernetes/resource.rb [hunk 2]
```diff
@@ -137,9 +137,12 @@ def update
       end
 
       def ensure_not_updating_match_labels
-        # blue-green deply is allowed to do this, see template_filler.rb + deploy_executor.rb
+        # blue-green deploy is allowed to do this, see template_filler.rb + deploy_executor.rb
         return if @template.dig(:spec, :selector, :matchLabels, :blue_green)
 
+        # allow manual migration when user is aware of the issue and wants to do manual cleanup
+        return if @template.dig(:metadata, :annotations, :"samson/allow_updating_match_labels") == "true"
```
Comment (adammw) [inline plugins/kubernetes/app/models/kubernetes/resource.rb]: do you have somewhere where you're putting all these annotations as constants? would be good to know what samson is looking for in one place.

**Review by adammw — APPROVED** [review]

File: plugins/kubernetes/app/models/kubernetes/resource.rb (hunk 2 above)
Comment (grosser) [reply plugins/kubernetes/app/models/kubernetes/resource.rb]: Readme has all the possible annotations :)

**Comment by uthark** [comment]
If users use apps/v1 api group, they can't update matchLabels:

> Note: In API version apps/v1, a Deployment’s label selector is immutable after it gets created.

https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#label-selector-updates

**Comment by grosser** [comment]
hoho, good thing we use extensions/v1beta1 everywhere :trollface:

**Comment by grosser** [comment]
... but still, I want to add an escape-hatch for this usecase, if it's not changeable in kubernetes then we need a downtime anyway

**Pull request merged**